pkg/operator/mon/health.go

```diff
@@ -143,9 +143,64 @@ func (c *Cluster) checkHealth() error {
 		return nil
 	}
 
+	// check if there are more than two mons running on the same node, failover one mon in that case
+	done, err := c.checkMonsOnSameNode()
+	if done || err != nil {
+		return err
+	}
+
+	done, err = c.checkMonsOnValidNodes()
+	if done || err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (c *Cluster) checkMonsOnSameNode() (bool, error) {
+	nodesUsed := map[string]struct{}{}
+	for name, node := range c.mapping.Node {
+		// when the node is already in the list we have more than one mon on that node
+		if _, ok := nodesUsed[node.Name]; ok {
+			// get list of available nodes for mons
+			availableNodes, _, err := c.getAvailableMonNodes()
+			if err != nil {
+				return true, fmt.Errorf("failed to get available mon nodes. %+v", err)
+			}
+			// if there are enough nodes for mons failover the mon too much from the node
+			if len(availableNodes) > 0 {
+				logger.Infof("rebalance: enough nodes available %d to failover mon %s", len(availableNodes), name)
+				c.failMon(len(c.clusterInfo.Monitors), name)
+			} else {
+				logger.Warningf("rebalance: not enough nodes available to failover mon %s", name)
+			}
+
+			// deal with one mon too much on a node at a time
+			return true, nil
```

Comment:
perhaps this return statement should be done when there are not available nodes to fail over. IE inside the previous `else` block .This will fix the case where mon_in_a_node=3 and available_nodes=2. So that this `for` loop runs again and fails over the other monitor

Reply:
by design only one failover happens per health check. Health checks are every 20s so a retry will be soon for another mon

Reply:
Even though the `failoverMon()` func blocks until the new ReplicaSet is created, those 20s allow the mon to "get in quorum" and in the worst case of the new mon failing to run the quorum checks before that should handle such issues directly.
Overall the two new checks are only run when cluster is in "full health" (/ all mons in quorum).